Fix controller mapping for Pioneer DDJ-SB

tonell-m opened the pull request:
Hi all,

I'm using Mixxx with a Pioneer DDJ-SB but the built in controller mapping was not fully working properly so I've fixed the following issues in this PR:

- When clicking the rotary selector to preview a track an error message was showing up because the script was missing a `rotarySelectorClick` handler.
  - Clicking the encoder now loads and starts the selected track in the preview deck. 
  - Clicking it a second time without changing the selected track will pause playback.
  - Preview playback now starts by default in the middle of the track similarly to DDJ-SB2 behaviour. The playback start position can be edited in the script's user options.
- The FX selection buttons were not working (clicking them did nothing), so the FX knobs could only control the Mix knob of each unit
  - For that I've removed the existing code for handling the FX units and replaced that with an implementation using the Components library

I've tested this updated script on Mixxx 2.3, 2.4 and 2.5 (all built from source) on macOS 12.6.7 (Monterey) using my DDJ-SB-L and it all seems to be working smoothly.

This is my first time contributing so let me know if some things are missing / need to be changed 🙂

Commit:
Fixed rotary encoder click behaviour for Pioneer DDJ-SB

- Added missing PioneerDDJSB.rotarySelectorClick function and made it start/stop preview deck for the selected track
- Playback start is now defaulted to the middle of the track, the playback start point is configurable via the script's user options section

Commit:
Codestyle cleaning

- Changed declaration of root object from `function() {}` to `{}`
- Removed spaces after `function` keyword

Commit:
Fixed FX controls

- Removed existing code for controlling effect units
- Replaced with implementation using Components

JoergAtGithub commented:
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

Swiftb0y commented on res/controllers/Pioneer-DDJ-SB.midi.xml:
```diff
@@ -236,15 +236,6 @@
                 </options>
             </control>
-            <control>
-                <group>[PreviewDeck1]</group>
-                <key>play</key>
-                <status>0x96</status>
-                <midino>0x66</midino>
-                <options>
-                    <normal/>
-                </options>
```
what did this button previously do?

tonell-m replied:
This button used to play the selected track in the preview deck, but because there was no custom behaviour to pause the preview when clicking the button a second time the only way to stop the preview was to stop it manually by clicking the pause button of the preview deck.

Swiftb0y replied:
Right, sorry I phrased my question wrong, what does the button do now? It seems like it doesn't do anything?

tonell-m replied:
Ah my bad, I didn't understand it that way. Now this behaviour is handled by the `rotarySelectorClick` midi mapping like so:

```xml
            <control>
                <group>[PreviewDeck1]</group>
                <key>LoadSelectedTrackAndPlay</key>
                <key>PioneerDDJSB.rotarySelectorClick</key>
                <status>0x96</status>
                <midino>0x41</midino>
                <options>
                    <normal/>
                    <script-binding />
                </options>
            </control>
```

(lines 745-753 of [Pioneer-DDJ-SB.midi.xml](https://github.com/mixxxdj/mixxx/pull/12280/files#diff-a94fec3379678bb49e2e22e882e73cba73a94b6b548a05f6c56969604806808f))

Swiftb0y replied:
Right, its good to know that no functionality was lost thanks to it being implemented in JS, but what I'm really interested in is whether the button is now unused. I can't spot it being handled anywhere so it just dead? What button on the controller is this exactly? I can't really guess just by looking at pictures.

If the button is completely unused now, that seems like a waste of valuable control-surface. Does that make sense?

tonell-m replied:
You're right I was missing something. According the [MIDI mappings sheet](https://www.pioneerdj.com/-/media/pioneerdj/software-info/controller/ddj-sb/ddj-sb_list_of_midi_messages_e.pdf) this button is actually the "BACK" button below the browser rotary selector. Thing is according to the manual its usage is to go back from browsing tracks into a crate/playlist to browsing the side panel. But in Mixxx browsing this panel is done using shift+rotary selector so this feature doesn't really have a proper usage. Do you have an idea of what relevant action could be done with this button instead?

Looking at the manual the shift+click action for this button is switch between panels in Serato DJ, so I'm thinking it could be used to toggle between showing/hiding the decks in Mixxx (what happens when you press the spacebar on your keyboard) but I'm not sure how to implement this, could you please guide me a bit?

<img width="442" alt="image" src="https://github.com/mixxxdj/mixxx/assets/19548672/d62b7e92-d0c0-4cbb-a9d8-bb2d6cb93d6f">

ronso0 replied:
In general it's recommended to make the buttons works as labeled. No need to reproduce special / unintuitive behaviour of other apps.
* "Back" would be [[Library],MoveFocusForward](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls.html#control-[Library]-MoveFocusForward), equivalent to pressing Tab on the keyboard. (Shift +press would be ..Backward (Shift + Tab)).
This allows to move focus between sidabar, tracks table and searchbar (and its Clear button)[1]. Focussing the search bar becomes relevant in Mixxx 2.4 where the searchbar is turned into a combobox that holds previous search queries.
* Rotating the encoder ([[Library],MoveVertical](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls.html#control-[Library]-MoveVertical)) would then allow selecting a query, top-most item is 'clear'.
* Pressing the rotary should be [`[Library],GoToItem`](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls.html#control-[Library]-GoToItem)
* Shift + press could be whatever you like (preview selected track). You may also implement Shift + longpress to have one more action (maximize library)

[1] we should discuss if we skip the Clear button in case a MoveFocus controls is used, unrelated to this PR

tonell-m replied:
Thanks for the feedback, I'll work on this. Although the alternative action for the BACK button is labelled as PANEL (see picture) so I think it would make sense for it to control toggling UI components when shift clicking, what do you think?

Swiftb0y replied:
maybe shift+back could be [`[Skin],show_maximized_library`](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls#control-[Master]-maximize_library) (we renamed it, the manual is outdated in that in that regard).

ronso0 replied:
Sure, no objections.
Use `[Skin],show_maximized_library` if that's what you need, and maybe mention in the manual that this control can be customized?

Swiftb0y commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -28,12 +28,22 @@ PioneerDDJSB.jogwheelSensitivity = 1.0;
+PioneerDDJSB.init = function(id) {
+
+    PioneerDDJSB.effectUnit = [];
+    PioneerDDJSB.effectUnit[1] = new PioneerDDJSB.EffectUnit(1);
+    PioneerDDJSB.effectUnit[2] = new PioneerDDJSB.EffectUnit(2);
```
wouldn't the (components.EffectUnit](https://github.com/mixxxdj/mixxx/wiki/Components%20JS#effectunit) suffice?

tonell-m replied:
I wasn't aware of that component, let me dig into it!

Swiftb0y replied:
Ah yeah, that makes sense. then the custom implementation is probably the better choice, I'll have a look asap.

tonell-m commented:
Done!

Swiftb0y commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
In what Mixxx version is this mapping broken exactly?

tonell-m replied:
Currently it is broken in all Mixxx versions I've tested (2.3, 2.4 and 2.5)

Swiftb0y replied:
Okay, so can you rebase this PR on 2.3 or 2.4? otherwise if you target `main` it'll be released with 2.5 the earliest.

Swiftb0y reviewed (commented):
Hey there, thank you for caring about this mapping. It is very much appreciated. Please see my questions down below. ~~Also, please understand that in order for us to merge this, we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.~~ (comment race :sweat_smile:)

Thank you.

Swiftb0y commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -898,88 +840,81 @@ PioneerDDJSB.rotarySelectorShiftedClick = function(channel, control, value, _sta
             }
+        },
+        outKey: "focused_effect",
+        output: function(value, _group, _control) {
+            this.send((value === this.buttonNumber) ? this.on : this.off);
+        },
```
fyi, this could be handled more elegantly via outValueScale:
```suggestion
        outValueScale: function(value) {
            return (value === this.buttonNumber) ? this.on : this.off;
        },
```

tonell-m replied:
Solved in https://github.com/mixxxdj/mixxx/pull/12280/commits/0406cbb5a95fe959399f29c22fe8a974229b6cad

Swiftb0y commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -898,88 +840,81 @@ PioneerDDJSB.rotarySelectorShiftedClick = function(channel, control, value, _sta
+    }
+
+    this.knob = new components.Pot({
+        unshift: function() {
+            this.input = function(channel, control, value, _status) {
+                value = (this.MSB << 7) + value;
+
+                var focusedEffect = engine.getValue(eu.group, "focused_effect");
+                if (focusedEffect === 0) {
+                    engine.setParameter(eu.group, "mix", value / this.max);
+                } else {
+                    var effectGroup = "[EffectRack1_EffectUnit" + unitNumber + "_Effect" + focusedEffect + "]";
+                    engine.setParameter(effectGroup, "meta", value / this.max);
+                }
+            };
+        },
```
there is also `inSetParameter` to simplify this a bit:
```suggestion
          inSetParameter: function(channel, control, value, _status) {
              if (engine.getValue(eu.group, "focused_effect") === 0) {
                  engine.setParameter(eu.group, "mix", value);
              } else {
                  var effectGroup = "[EffectRack1_EffectUnit" + unitNumber + "_Effect" + focusedEffect + "]";
                  engine.setParameter(effectGroup, "meta", value);
              }
          },
```
and if you don't overwrite anything in the `shift` method, you also don't need to touch the `unshift`.

tonell-m replied:
Solved in https://github.com/mixxxdj/mixxx/pull/12280/commits/00d56cce043849447b7d7620cffe125b4aea330d

Holzhaus commented:
Thanks for your contribution. I changed the base branch to 2.4, as we won't do another 2.3 release. Could you please check if the mapping still matches the behavior documented in manual (https://manual.mixxx.org/2.4/en/hardware/controllers/pioneer_ddj_sb) and file a PR for the manual repository if changes are needed?

Commit:
Default preview start to 30% of the track instead of 50, according to current documentation

Commit:
Use outValueScale instead of output for handling focused effect

Commit:
Simplify code by using inSetParameter

Instead of overriding unshift behaviour with a custom input function

tonell-m commented:
Hi, I'm having some trouble with my Components implementation and I would please need some help with it. I'm trying to make a custom Button implementation for my effect buttons so I have something like this: 

```js
PioneerDDJSB.EffectUnit = function(unitNumber) {
    components.ComponentContainer.call(this);

    this.EffectButton = function(number, options) {
        this.number = number;
        this.group = eu.group;
        this.midi = [0x93 + unitNumber, 0x46 + number];

        components.Button.call(this);
    };
    this.EffectButton.prototype = new components.Button({
        input: ...,
        outKey: "focused_effect",
        output: ...
    });

    this.buttons = new components.ComponentContainer();
    for (var i = 1; i <= 3; i++) {
        this.buttons[i] = new this.EffectButton(i);
        ...
    }
};
PioneerDDJSB.EffectUnit.prototype = new components.ComponentContainer();
```

Now my issue is that I get an error when my script is loaded saying:
```
Uncaught exception: file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:73: TypeError: Property 'outValueScale' of object [object Object] is not a function
Backtrace: output@file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:73
@file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:140
trigger@file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:139
Component@file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:52
Button@file:///Users/tonell_m/Dev/mixxx/res/controllers/midi-components-0.0.js:165
@file:///Users/tonell_m/Dev/mixxx/res/controllers/Pioneer-DDJ-SB-scripts.js:952
@file:///Users/tonell_m/Dev/mixxx/res/controllers/Pioneer-DDJ-SB-scripts.js:987
@file:///Users/tonell_m/Dev/mixxx/res/controllers/Pioneer-DDJ-SB-scripts.js:55"
```

The line in my script that triggers this error is `components.Button.call(this);`, that gets called when creating the EffectUnit instance from the `init` function of `PioneerDDJSB`. I've tried logging `this` in the console but the object that is being printed is the top level controller object of my script (`PioneerDDJSB`) and not my `EffectButton` instance as I would expect. I'm sure I'm missing something there but I'm not sure what, as I'm far from being a JS expert. Do you have any idea of what is going wrong there? Any help would be greatly appreciated ❤️

Commit:
Update controller FX buttons to enable/disable effects according to manual

- The DDJ-SB manual states that a non shift press on the buttons should toggle effects on or off instead of selecting the focused effect (this should be done by shift clicking the buttons as the controller labels indicate)
- Renamed `buttons` to `enableButtons` for the sake of clarity and consistency with the default EffectUnit implementation

Commit:
Updated usages of deprecated functions and parameter names

- Updated parameter names that were marked as deprecated
- Updated usage of `connectControl` in favour of `makeConnection` using a new utility method in the common controller script to limit modifications and allow easier update of other DDJ-SB* scripts

Commit:
Updated rotary selector controls

- Turning rotary selector moves vertically in the Library using [Library] MoveVertical
- Clicking rotary selector calls [Library] GoToItem

Commit:
Use shift + rotary encoder to move playback position in preview deck

Since this control is now unused and shift + click loads the track in the preview deck, it can be useful to move the playhead around to quickly preview different sections of the selected track

Commit:
Shift + FX button click now allows to select currently focused effect

- Toggling shift updates the fx buttons leds to show either enabled effects (unshift) or currently focused effect (shift)
- Bound right shift button to also trigger `PioneerDDJSB.shiftButton`

Commit:
Updated BACK button behaviour

- Clicking BACK calls [Library] MoveFocusBackward to allow switching focused library element
- Shift clicking BACK maximises the library by toggling [Skin] show_max_library

tonell-m commented:
@Swiftb0y @ronso0 I've updated my implementation to take all your comments into consideration, I would recommend you look at the final result instead of going commit per commit as there has been a bit of trial and error 😅 

The final mappings are the following:
- Browser:
  - Turning rotary selector: `[Library] MoveVertical`
  - Clicking rotary selector: `[Library] GoToItem`
  - Shift clicking rotary selector: `[PreviewDeck1] LoadSelectedTrackAndPlay`
  - Shift turning rotary selector: moves `[PreviewDeck1] playposition` around to allow scrolling through the previewed track
  - Clicking BACK button: `[Library] MoveFocusBackward` (I think it makes more sense than `MoveFocusForward` as it directly selects the tree view when the focus is on the track list, which is usually what we want)
  - Shift clicking BACK button: toggles `[Skin] show_maximized_library` (as the BACK button shift label on the controller is "PANNEL")
- Effect units:
  - Clicking FX buttons: toggles `[EffectRack1_EffectUnitX_EffectX] enabled`
  - Shift clicking FX buttons: changes `[EffectRack1_EffectUnitX] focused_effect`
  - Holding the shift button updates the FX buttons LEDs to show the currently selected effect

If you're okay with these mappings I'll submit a PR for updating manual to reflect those changes. Let me know if you have further comments / questions!

daschuer commented:
Oh sorry to not review this in time. Now I see a conflict has developed. 
@Swiftb0y how is the state here?

acolombier commented:
@tonell-m do you have interest to continue this PR? As I understand, it only need to get the conflict fixed to be merged. Happy to help with that otherwise.

tonell-m commented:
@acolombier sorry for the late response, I haven't had much time to continue this PR lately and probably won't either in upcoming weeks, so if you want to pick it up from there, feel free to do so!
Otherwise I will come back to it later :)

acolombier commented:
No problem at all - glad to hear you haven't lost interest to complete it at some point! I'll mark it as draft for now, but feel free to mark ready as soon as we should get back to it

Commit:
Merge remote-tracking branch 'upstream/2.5' into fix/pioneer-ddj-sb-mapping

Commit:
Fix updateLeds function when clicking button before pressing shift

Commit:
Remove lodash dependency

ronso0 commented:
FYI in order to prepare for the 2.6 release (and clean up our milestone list) I just moved this to the 2.6.1 milestone and changed the base to 2.6
